### PR TITLE
feat(codux-ui): create new package codux-ui

### PR DESCRIPTION
**Pull request by 0xRezaa**
This PR creates a new package in codux-core called codux-ui (@wixc3/codux-ui).
It will contain UI components which integrate well with codux.

WixImage:
For the beginning it rolls out with a image component which handles wix media images and handles responsive design.
Through setting the `format wix-image-id` the component integrates with codux. In codux we know to show the wix-media-manager. Once the user has selected an image we write the image id in our format into the code (we try to stick to the definition of wix internal identifiers defined [here](https://dev.wix.com/docs/sdk/core-modules/sdk/media)).
src and image id relate like following
src: `https://static.wixstatic.com/media/123456/v1/fill/w_20/anyName.jpg`
imageId: `wix:image://v1/123456/anyName.jpg`

### CODE REVIEW

File: packages/codux-ui/src/image/index.ts
Comment (oleksandraab) [inline packages/codux-ui/src/image/index.ts:1]: Is this empty file needed?

File: packages/codux-ui/src/index.ts
Comment (oleksandraab) [inline packages/codux-ui/src/index.ts:1]: Is this empty file needed?

File: packages/codux-ui/src/image/image.tsx [hunk 1]
```diff
@@ -0,0 +1,120 @@
+import React from 'react';
+import { createBoard } from '@wixc3/react-board';
+import { WIX_MEDIA_ID_ROOT, WIX_MEDIA_ROOT } from './constants';
+
+export default createBoard({
```
Comment (barak007) [inline packages/codux-ui/src/image/image.tsx:5]: I feel like the board is not part of the component.

File: packages/codux-ui/src/image/image.tsx [hunk 2]
```diff
@@ -0,0 +1,120 @@
+import React from 'react';
```
Comment (barak007) [inline packages/codux-ui/src/image/image.tsx:1]: to my eyes this file main export is the Image component. I would put it at the top and try to keep this file as small as possible.

Comment (barak007) [reply packages/codux-ui/src/image/image.tsx:1]: it means that utils and types should go outside.

File: packages/codux-ui/package.json [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      "import": "./dist/image/index.js",
+      "require": "./dist/image/index.js"
+    }
+  },
+  "peerDependencies": {
+    "react": ">=18.0.0 || ^19.0.0-rc.0",
+    "react-dom": ">=18.0.0 || ^19.0.0-rc.0"
+  },
+  "dependencies": {
+    "@wixc3/react-board": "^4.3.3"
```
Comment (barak007) [inline packages/codux-ui/package.json:21]: I feel like this should be a dev dependency.

File: packages/codux-ui/package.json [hunk 4]
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "@wixc3/codux-ui",
+  "description": "Common ui components for codux",
+  "version": "4.5.1",
```
Comment (barak007) [inline packages/codux-ui/package.json:4]: do we release all versions of this monorepo together? should the versions be synced?

File: packages/codux-ui/README.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+#### Example for the `ResponsiveImage` component:
+
+```tsx
+import { ResponsiveImage } from '@wixc3/codux-ui/image';
+
+<ResponsiveImage
+  alt="Sample image"
+  imageId="wix:image://v1/11062b_96503e81a83e47ed857a44be26ebd0d1~mv2.jpeg/Rustic breakfast.jpeg"
+  mediaBreakpoints={[
```
Comment (barak007) [inline packages/codux-ui/README.md:33]: using objects as props causes this component to over render.
maybe the example should put this config outside the component, or use a string to specify mediaBreakpoints. it's kind of the ugly side of react.

Comment (barak007) [reply packages/codux-ui/README.md:33]: maybe we could also create a type that validates this string.

File: packages/codux-ui/src/image/index.ts
Comment (barak007) [reply packages/codux-ui/src/image/index.ts:1]: remove index.

File: packages/codux-ui/src/index.ts
Comment (barak007) [reply packages/codux-ui/src/index.ts:1]: remove index.

File: packages/codux-ui/package.json [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  "main": "dist/index.js",
+  "exports": {
+    "./test-kit": {
+      "import": "./dist/test-kit/index.js",
```
Comment (barak007) [inline packages/codux-ui/package.json:8]: I don't see a test-kit here. remove exports.

File: packages/codux-ui/README.md (hunk 5 above)
Comment (barak007) [reply packages/codux-ui/README.md:33]: maybe a custom memo on the component? we need to think about it a bit.

File: packages/codux-ui/src/image/image.tsx [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+}: WixImageAttributes & { imageId: string }) {
+    const { wixMediaId, fileName } = extractWixMediaIdAndTitle(imageId);
+    const { title, ext } = extractFileExtension(fileName);
+    const encodedFileName = `${encodeURIComponent(displayName || title)}.${ext || '0'}`;
+    const srcString = `${WIX_MEDIA_ROOT}${wixMediaId}/v1/${renderingStrategy}/w_${width},h_${height}/${encodedFileName}`;
+
+    return srcString;
+}
+
+export const WixImage: React.FC<WixImageProps> = ({ className, imageId, mediaBreakpoints, alt }) => {
```
Comment (barak007) [inline packages/codux-ui/src/image/image.tsx:100]: try to compare manual picture element usage (with some util for url generation) vs this Image component.
I want to evaluate if the abstraction worth it.

**Review by barak007 — COMMENTED** [review]
I did a quick look. let's talk after another pass here.

File: packages/codux-ui/package.json [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+      "require": "./dist/test-kit/index.js"
+    },
+    "./image": {
+      "import": "./dist/image/index.js",
```
Comment (AviVahl) [inline packages/codux-ui/package.json:12]: this and the next line make no sense together.

File: packages/codux-ui/src/image/image.tsx [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        />
+    ),
+    environmentProps: {
+        windowWidth: 820,
+        windowHeight: 1180,
+    },
+});
+
+interface WixImageAttributes {
+    displayName?: string;
```
Comment (AviVahl) [inline packages/codux-ui/src/image/image.tsx:34]: export all interfaces.
add jsdocs

File: packages/codux-ui/src/image/image.tsx [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    renderingStrategy: 'fit' | 'fill';
+}
+
+interface MediaBreakPoint {
+    /** @important */
+    minWidth: number;
+}
+
+export interface WixImageProps {
+    /** @format wix-image-id */
```
Comment (AviVahl) [inline packages/codux-ui/src/image/image.tsx:46]: ?

File: packages/codux-ui/src/image/image.tsx [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+            ]}
+        />
+    ),
+    environmentProps: {
+        windowWidth: 820,
+        windowHeight: 1180,
+    },
+});
+
+interface WixImageAttributes {
```
Comment (AviVahl) [inline packages/codux-ui/src/image/image.tsx:33]: export all interfaces.
add jsdocs

**Review by AviVahl — COMMENTED** [review]
Media platform. Do they have a typed public SDK?

File: packages/codux-ui/package.json [hunk 12]
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "@wixc3/codux-ui",
+  "description": "Common ui components for codux",
```
Comment (AviVahl) [inline packages/codux-ui/package.json:3]: "for Codux"?

File: packages/codux-ui/package.json (hunk 4 above)
Comment (0xRezaa) [reply packages/codux-ui/package.json:4]: I am not sure. It seems that some versions differ. I synched them with app-core
@PeterShershov wdyt?

File: packages/codux-ui/src/image/image.tsx (hunk 7 above)
Comment (0xRezaa) [reply packages/codux-ui/src/image/image.tsx:100]: The point is that through adding the `@format wix-media-id` to the jsdoc of the imageId property we have integration in codux with our wix-media-id visualizer.

It basically allows us to open the wix-media-manager to pick a image and sets the id as the property. The properties are configured by the user directly

**Comment by 0xRezaa** [comment]
> Media platform- do they have a typed public SDK?

Afaik no. There are utils in platform repos but not publicly published.

File: packages/codux-ui/src/image/image.tsx [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+    displayName?: string;
+    height: number;
+    width: number;
+    renderingStrategy: 'fit' | 'fill';
+}
+
+interface MediaBreakPoint {
+    /** @important */
```
[the remark this answers is not in the document]
Comment (0xRezaa) [reply packages/codux-ui/src/image/image.tsx:41]: This is part of the codux integration. When drilling into the object visualizer of an element of the mediaBreakPoints property, we want to see the `minWidth` on top of that list. @important tells codux how to sort the properties of that object.

File: packages/codux-ui/README.md (hunk 5 above)
Comment (0xRezaa) [reply packages/codux-ui/README.md:33]: Good point. The problem is not the validation but the integration with existing visualizer.
So I think a custom memo would make sense

File: packages/codux-ui/package.json (hunk 3 above)
Comment (0xRezaa) [reply packages/codux-ui/package.json:21]: Was removed

File: packages/codux-ui/src/image/image.tsx (hunk 1 above)
Comment (0xRezaa) [reply packages/codux-ui/src/image/image.tsx]: exactly. It was removed

File: packages/codux-ui/src/image/image.tsx (hunk 2 above)
Comment (barak007) [reply packages/codux-ui/src/image/image.tsx:1]: I think only `createWixImageUrlFromWixImageAttributes` should be exported but I would change the name. too long.

**Comment by barak007** [comment]
The one thing that is missing and also a must, is some tests to make sure we don't break it each release. also CI will not feel lonely.
You want to have unit test to url creation and some tests for the component. we could strait up using playwright tests.

File: packages/codux-ui/README.md (hunk 5 above)
Comment (barak007) [reply packages/codux-ui/README.md:33]: are we going for it?

**Comment by barak007** [comment]
One more thing to think about is to add native html attributes and allow to pass events and aria. usually it's a good practice for robust components.

**Comment by 0xRezaa** [comment]
@barak007 I agree with both suggestions

File: packages/codux-ui/src/image/image.tsx (hunk 2 above)
Comment (0xRezaa) [reply packages/codux-ui/src/image/image.tsx:1]: I renamed it to buildWixImageUrl and exported it

**Comment by 0xRezaa** [comment]
> ust, is some tests to make sure we don't break it each release. also CI will not feel lonely.

I added tests for the utils to build url's and some type check cases for the component.
@AviVahl  To make the test setup work I set in the tsconfig of codux-ui/test `"module": "Preserve"` and `"moduleResolution": "bundler"`. I don't see any reason why this should be a problem but please clarify if I got something wrong.

File: packages/codux-ui/README.md (hunk 5 above)
Comment (0xRezaa) [reply packages/codux-ui/README.md:33]: Yes. I added it

File: packages/codux-ui/src/image/image.tsx (hunk 11 above)
Comment (0xRezaa) [reply packages/codux-ui/src/image/image.tsx]: added

**Review by oleksandraab — APPROVED** [review]

**Comment by 0xRezaa** [comment]
@AviVahl As discussed we set:
- `"type": "module"` in the package.json of the package setting it to esm only
- import extensions are set to `.js`to make it node compatible with the updated module and moduleResolution compilerOptions in the tsconfig base
- picture and source setup is omitted in favor of pure img with srcSet
- specific compiler options for the tests are removed

**Comment by 0xRezaa** [comment]
We refactored back from a pure `img` with `srcset` to the `picture` with `source` element setup.
This is due to more gradual control with the picture setup about which source to chose based on a media condition.
In case of a srcset the choice of the src is automatic happening based on pixel density and viewport size.
Thus we pass imgProps onto the img component.
We chose to expose the component as `WixImage`still as it sounds more natural and familiar to non technical users

**Review by barak007 — APPROVED** [review]

File: packages/codux-ui/test-kit/render.tsx [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+const ResolveOnMount = ({ children, onMount }: React.PropsWithChildren<{ onMount: () => void }>) => {
+    useEffect(onMount);
+    return <>{children}</>;
+};
+
+/**
+ * TODO(global dispose): This is here until "render" will accept disposables.
+ */
+const disposables = createTestDisposables();
```
[the remark this answers is not in the document]
Comment (0xRezaa) [reply packages/codux-ui/test-kit/render.tsx]: Fixed

**Pull request merged**